File: src/main/java/dev/slimevr/posestreamer/TickPoseStreamer.java
package dev.slimevr.posestreamer;

import dev.slimevr.vr.processor.skeleton.Skeleton;

import java.io.IOException;


public class TickPoseStreamer extends PoseStreamer {

	protected long nextFrameTimeMs = -1L;

	public TickPoseStreamer(Skeleton skeleton) {
		super(skeleton);
	}

	public void doTick() {
		PoseDataStream poseFileStream = this.poseFileStream;
		if (poseFileStream == null) {
			return;
		}

		Skeleton skeleton = this.skeleton;
		if (skeleton == null) {
			return;
		}

		long curTime = System.currentTimeMillis();
		if (curTime < nextFrameTimeMs) {
			return;
		}

		nextFrameTimeMs += frameRecordingInterval;

		// To prevent duplicate frames, make sure the frame time is always in
		// the future
		if (nextFrameTimeMs <= curTime) {
			nextFrameTimeMs = curTime + frameRecordingInterval;
		}

		captureFrame();
	}

	@Override
	public synchronized void setOutput(PoseDataStream poseFileStream) throws IOException {
		super.setOutput(poseFileStream);
		nextFrameTimeMs = -1L; // Reset the frame timing
	}
}

File: server/core/src/main/java/dev/slimevr/posestreamer/TickPoseStreamer.kt
package dev.slimevr.posestreamer

import dev.slimevr.tracking.processor.skeleton.HumanSkeleton
import dev.slimevr.util.TickReducer
import java.io.IOException

open class TickPoseStreamer(skeleton: HumanSkeleton) : PoseStreamer(skeleton) {
	private val ticker = TickReducer({ captureFrame() }, frameInterval)

	@set:Throws(IOException::class)
	@set:Synchronized
	override var output: PoseDataStream?
		get() = super.output
		set(value) {
			super.output = value
			ticker.reset()
		}

	@set:Synchronized
	override var frameInterval: Float
		get() = super.frameInterval
		set(value) {
			super.frameInterval = value
			ticker.interval = value
		}

	fun tick(tickDelta: Float) {
		// Only tick if there is an output
		if (hasOutput) {
			ticker.tick(tickDelta)
		}
	}
}
